### DISCUSSION TITLE
Using prefetch to improve message consumption latency/throughput

### DISCUSSION THREAD

**Discussion by andrzejolszak** [rabbitmq/rabbitmq-dotnet-client#1200] [Q&A] [answered]
Hello,

I'm looking at a scenario where I need to receive messages from a 3rd party publisher, and I'm looking for ways of minimizing/hiding latency.
I don't have any control over the publisher and its network stack, but can adjust the configuration on my side.

As the baseline, please consider this basic code:
```
var consumer = new EventingBasicConsumer(channel);
consumer.Received += OnMessageReceived;
channel.BasicConsume(queueName, autoAck: true, consumer: consumer);
```

My questions:
1. I believe no prefetching happens in the above example, as according to the documentation setting `autoAck=true`, makes the prefetching settings `prefetchCount` and `pretechSize` be ignored. Thus, the message transmission over the network will be completely sequential. Please correct me if I'm wrong?

2. If I enable prefetching (via explicit ACKs and `channel.BasicQos`), can this possibly make the messages arrive out of order from the point of view of the `OnMessageReceived` ?
If yes, can DeliveryTag or some other property be used to check the ordering and reliably fix it afterwards?

**Comment by michaelklishin** [comment, marked as answer]
You misunderstand what prefetching means. Prefetch does not affect delivery order but re-queueing can.

Prefetching simply controls the number of "in flight" deliveries that RabbitMQ would allow when manual acknowledgements are used.

See the guide on [Acknowledgements](https://rabbitmq.com/confirms.html).

**Reply by andrzejolszak** [reply]
Thanks for help, this makes sense now.

**Comment by michaelklishin** [comment]
So the setting controls how many deliveries can be transferred on the channel, and dispatched to an `OnMessageReceive` handler. Those dispatched messages *can* and in practice *will* be processed in parallel assuming the runtime has enough cores.
